server/controllers/secret.js

```diff
@@ -31,51 +31,60 @@ async function getSecretRoute(request, reply) {
         return reply.code(403).send({ error: 'Not a valid secret ID' });
     }
 
-    const data = await prisma.secret.findFirst({
-        where: {
-            id,
-            expiresAt: { gte: new Date() },
-        },
-        include: { files: true },
-    });
+    // Use a transaction to ensure atomicity of read and delete operations
+    const result = await prisma.$transaction(async (tx) => {
+        const data = await tx.secret.findFirst({
+            where: {
+                id,
+                expiresAt: { gte: new Date() },
+            },
+            include: { files: true },
+        });
```

Comment:
I don't think transaction alone enables read-side locking. If I understand correctly, the race condition may still exist.
See https://github.com/prisma/prisma/issues/17136
See https://github.com/prisma/prisma/issues/17136#issuecomment-2272195564 for workaround

Until natively supported with the `find*` methods, you'll likely need to use reach for `$queryRaw` to get `lock for update`

Reply:
Interesting. I will look into this. Did not expect anyone to review, glad you did!